File: Grind75/73-MaxProfitJobScheduling/cpp/MaxProfitJobScheduling.cpp
#include "MaxProfitJobScheduling.h"
#include <algorithm>
#include <cstdint>
#include <cstdio>
#include <iterator>
#include <sys/types.h>
#include <tuple>
#include <vector>

using namespace std;

int Solution::schedule(vector<tuple<int, int, int>> &jobs, uint index, int lastEnd) {
  if (index >= jobs.size())
    return 0;
  const auto &[start, end, profit] = jobs[index];
  int skip = schedule(jobs, index + 1, lastEnd);
  if (start < lastEnd)
    return skip;
  int take = profit + schedule(jobs, index + 1, end);
  return max(take, skip);
}

int Solution::jobScheduling2(
    vector<int>& startTime,
    vector<int>& endTime,
    vector<int>& profit) {
  vector<tuple<int, int, int>> jobs;
  for (uint i = 0; i < startTime.size(); i++)
    jobs.push_back({startTime[i], endTime[i], profit[i]});
  std::sort(jobs.begin(), jobs.end());
  return schedule(jobs, 0, 0);
}

int Solution::endBeforeStart(int index, vector<tuple<int, int, int>> &jobs) {
  const int &start = get<1>(jobs[index]);
  int left = 0, right = index - 1;
  while (left <= right) {
    int center = (left + right) / 2;
    const int &end = get<0>(jobs[center]);
    if (end < start) {
      left = center + 1;
    } else if (end > start) {
      right = center - 1;
    } else {
      right = center;
      break;
    }
  }
  int best = right;
  while (best + 1 < index && get<0>(jobs[best+1]) <= start)
    best++;
  return best;
}

int Solution::jobScheduling3(
    vector<int>& startTime,
    vector<int>& endTime,
    vector<int>& profit) {
  if (startTime.empty())
    return 0;

  vector<tuple<int, int, int>> jobs;
  for (uint i = 0; i < startTime.size(); i++)
    jobs.push_back({endTime[i], startTime[i], profit[i]});
  std::sort(jobs.begin(), jobs.end());

  vector<int> bestProfit(jobs.size(), 0);
  for (uint j = 0; j < jobs.size(); j++) {
    const auto &[end, start, profit] = jobs[j];
    const int lastEnd = endBeforeStart(j, jobs);
    const int lastBest = lastEnd >= 0 ? bestProfit[lastEnd] : 0;
    const int prev = j > 0 ? bestProfit[j-1] : 0;
    bestProfit[j] = max(prev, lastBest + profit);
  }
  return bestProfit.back();
}


int Solution::jobScheduling(
    vector<int>& startTime,
    vector<int>& endTime,
    vector<int>& profit) {
  if (startTime.empty())
    return 0;

  vector<tuple<int, int, int>> jobs;
  for (uint i = 0; i < startTime.size(); i++)
    jobs.push_back({endTime[i], startTime[i], profit[i]});
  sort(jobs.begin(), jobs.end());

  vector<int> cache(jobs.size(), 0);
  for (uint j = 0; j < jobs.size(); j++) {
    const auto &[end, start, profit] = jobs[j];
    const tuple<int, int, int> search {start, INT32_MAX, INT32_MAX};
    const int lastEnd =
      upper_bound(jobs.begin(), jobs.end(), search) - jobs.begin() - 1;
    const int lastBest = lastEnd >= 0 ? cache[lastEnd] : 0;
    const int prev = j > 0 ? cache[j-1] : 0;
    cache[j] = max(prev, lastBest + profit);
  }
  return cache.back();
}
 

File: Grind75/73-MaxProfitJobScheduling/python/MaxProfitJobScheduling.py
#!/bin/python3

from __future__ import annotations
import sys
from typing import Tuple
sys.path.append('../../Utility/python')
from PrettyPrint import *
from Colors import *
import bisect
from math import inf

class Solution:
  def jobScheduling(self,
                    startTime: List[int],
                    endTime: List[int],
                    profit: List[int]) -> int:
    jobs: List[Tuple[int, int, int]] = sorted(zip(endTime, startTime, profit))
    cache: List[int] = [0] * (len(jobs))
    for j in range(len(jobs)):
      _,s,p = jobs[j]
      prevJob = bisect.bisect_right(jobs, (s, inf, inf)) - 1
      cache[j] = max(cache[j-1], p + cache[prevJob])
    return cache[-1] if len(cache) > 0 else 0

if __name__ == '__main__':
  soln = Solution()
  testCases = [
    ( [], [], [], 0 ),
    ( [1], [2], [3], 3 ),
    ( [1,2], [2,3], [3,4], 7 ),
    ( [1,1], [2,3], [3,4], 4 ),
    ( [1,1,2], [2,3,3], [3,4,5], 8 ),
    ( [1,2,3,4], [4,4,4,5], [1,3,2,4], 7 ),
    ( [1,2,3,4,6], [3,5,10,6,9], [20,20,100,70,60], 150 ),
  ]

  for start, end, profit, answer in testCases:
    # jobs = list(zip(start, end, profit))
    # jobs.sort(key=lambda j: j[1])
    # for j in range(len(jobs)):
    #   s, e, p = jobs[j]
    #   colorPrint(LIGHT_BLUE, f'{j}: ({s}->{e}, {p})')
    # for j in range(len(jobs)):
    #   print('prev disjoint of', j, 'is', soln._upperBound(jobs, j))
    colorPrint(LIGHT_BLUE, start)
    colorPrint(LIGHT_BLUE, end)
    colorPrint(LIGHT_BLUE, profit, '=>', answer)
    result = soln.jobScheduling(start, end, profit)
    colorPrint(GREEN if result == answer else RED, result)

